oeps/oep-0037-bp-test-data.rst

```diff
@@ -0,0 +1,299 @@
+desired test or development task.  For relational data in simple unit tests,
+this is often done in Python via the `factory_boy`_ package (and custom model
+factories using it) directly in ``setUpTestData`` or ``setUp``.  For
+development data (or more complex test setup which could be useful to
+reproduce in a development environment), the setup code should instead reside
+in its own function or static method.  For example:
+
+.. code-block:: python
+
+    def account_retirement_example(data):
+        data.test_user = UserFactory.create()
+        data.test_superuser = SuperuserFactory.create()
+        data.original_email = data.test_user.email
+
+        # Misc. setup
+        PendingEmailChangeFactory.create(user=data.test_user)
+        UserOrgTagFactory.create(user=data.test_user, key='foo', value='bar')
+        UserOrgTagFactory.create(user=data.test_user, key='cat', value='dog')
+
+        CourseEnrollmentAllowedFactory.create(email=data.original_email)
+
+(JavaScript tests seeking to stub out data returned from a server can use
+similar functions utilizing a package like `rosie`_ or `factory-girl`_.)
+
+Setting attributes of the provided argument like this allows the test object
+or class to be provided when used in a unit test (so assertions can be written
+naturally), and something like a `SimpleNamespace`_ object can be provided
+when run from a developer console so the generated data can be quickly located
+and inspected.  When needed, course data should be loaded from an `OLX`_
+archive via ``call_command()`` using the existing management commands for
+course import.
+
+Only the attributes needed to perform the intended task should be specified;
+the factory class should take care of setting reasonable defaults for the
+rest.  The `Faker`_ package can be used by factories to generate
+reasonable-looking values for many common types of data.  By auto-populating
+many of the fields this way, when a model changes often only the factory will
+need to be updated rather than every data set using that model.
+
+The function can take one or more positional arguments if useful for using it
+in different contexts.  It could for example create records linked to a
+provided existing user, or populate them based on information loaded from a
+YAML file found at the provided path.  Keyword-only arguments should usually
+be avoided, as they are more difficult to use from the kind of general-purpose
+management commands and ``make`` targets described below.
+
+.. _factory_boy: https://factoryboy.readthedocs.io/en/latest/index.html
+.. _rosie: https://www.npmjs.com/package/rosie
+.. _factory-girl: https://www.npmjs.com/package/factory-girl
+.. _SimpleNamespace: https://docs.python.org/3/library/types.html#types.SimpleNamespace
+.. _OLX: https://edx.readthedocs.io/projects/edx-open-learning-xml/en/latest/index.html
+.. _Faker: https://faker.readthedocs.io/en/stable/
+
+Loading Data
```

Comment:
This OEP is great! It would be nice if we explicitly dealt with the case where test data is loaded multiple times. Ex. into a dev environment, modified by hand, and then reloaded. I would expect that in that case the data loading function would stomp on the manual changes, but it would be good to have at least some guidance so people don't need to know the details of every command to know what will happen when it gets re-run.

Reply:
Because we're trying to avoid specifying primary keys, the most likely behavior is that reloading a fixture will add a new set of comparable records with different IDs.  Unless that poses a problem, I'll just document that behavior here.